Add division (div) operator to calc plugin

panzerdev opened the pull request:
Fix https://github.com/evcc-io/evcc/issues/5314

panzerdev committed:
Add division (div) operator to calc plugin 

PR for feature proposal https://github.com/evcc-io/evcc/issues/5314

panzerdev committed:
update error text on wrong calc operator

andig commented on provider/calc.go:
```diff
@@ -30,8 +31,8 @@ func NewCalcFromConfig(other map[string]interface{}) (IntProvider, error) {
 
-	if len(cc.Add) > 0 && len(cc.Mul) > 0 || len(cc.Add) > 0 && cc.Sign != nil || len(cc.Mul) > 0 && cc.Sign != nil {
-		return nil, errors.New("can only have either add, mul or sign")
+	if len(cc.Add) > 0 && len(cc.Mul) > 0 && len(cc.Div) > 0 || len(cc.Add) > 0 && cc.Sign != nil || len(cc.Mul) > 0 && cc.Sign != nil {
```
der Test funktioniert nicht

andig replied:
Mit ner kleinen Hilfsfunktion zur Vereinfachung, genau.

andig commented on provider/calc.go:
```diff
@@ -102,6 +111,21 @@ func (o *calcProvider) floatGetter() (float64, error) {
+			if err != nil {
+				return 0, fmt.Errorf("div[%d]: %w", idx, err)
+			}
+			if res == 0 {
```
was machst Du wenn res beim zweiten Operand immer noch Null ist? Die Logik erscheint fehlerhaft?

andig replied:
Stimmt :). Braucht aber einen kleinen Kommentar. Und Div/0 müsste vmtl auch abgefangen oder wenigstens mal getestet werden.

panzerdev replied:
If the dividend is 0 then the result is 0 right away, otherwise the value will be divided by every divisor != 0. Divisor == 0 are ignored and don't change the result

panzerdev committed:
Fix parameter validation in calc plugin

panzerdev committed:
improve logic on calc plugin division

If the dividend is 0 then the result is 0 right away, otherwise the value will be divided by every divisor != 0. Divisor == 0 are ignored and don't change the result

andig committed:
Update calc.go

andig commented on provider/calc.go:
```diff
@@ -102,6 +114,23 @@ func (o *calcProvider) floatGetter() (float64, error) {
+				} else {
+					res = v
+				}
+			} else if v != 0 {
```
was passiert bei v==0?

panzerdev replied:
v == 0 zu diesem Zeitpunkt wäre eine division durch 0 von daher würde ich das als noop machen und den Wert unverändert lassen

andig replied:
Und beim nächsten lustig weiter dividieren? Wie soll man den Fehler jemals finden?

panzerdev replied:
Einig sind wir uns ja bei 0 nicht weiter dividieren zu können. Was hälst von error log ausgabe und `res == 0`?

andig replied:
Gerne res 0, kein Log. Gibts an anderer Stelle auch nicht. Wenigstens kann dann nicht ein willkürlicher Wert raus kommen.

panzerdev committed:
Return 0 on division by 0
